internal/clients/api_client.go

```diff
@@ -617,3 +618,131 @@ func (a *ApiClient) GetElasticsearchSettings() (map[string]interface{}, diag.Dia
 	}
 	return clusterSettings, diags
 }
+
+func (a *ApiClient) PutElasticsearchIndex(index *models.Index) diag.Diagnostics {
+	var diags diag.Diagnostics
+	indexBytes, err := json.Marshal(index)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	log.Printf("[TRACE] index definition: %s", indexBytes)
+
+	req := a.es.Indices.Create.WithBody(bytes.NewReader(indexBytes))
+	res, err := a.es.Indices.Create(index.Name, req)
+	if err != nil {
+		diag.FromErr(err)
+	}
+	defer res.Body.Close()
+	if diags := utils.CheckError(res, fmt.Sprintf("Unable to create index: %s", index.Name)); diags.HasError() {
+		return diags
+	}
+	return diags
+}
+
+func (a *ApiClient) DeleteElasticsearchIndex(name string) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	res, err := a.es.Indices.Delete([]string{name})
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	defer res.Body.Close()
+	if diags := utils.CheckError(res, fmt.Sprintf("Unable to delete the index: %s", name)); diags.HasError() {
+		return diags
+	}
+
+	return diags
+}
+
+func (a *ApiClient) GetElasticsearchIndex(name string) (*models.Index, diag.Diagnostics) {
+	var diags diag.Diagnostics
+
+	req := a.es.Indices.Get.WithFlatSettings(true)
+	res, err := a.es.Indices.Get([]string{name}, req)
+	if err != nil {
+		return nil, diag.FromErr(err)
+	}
+	defer res.Body.Close()
+	// if there is no index found, return the empty struct, which should force the creation of the index
+	if res.StatusCode == http.StatusNotFound {
+		return nil, nil
+	}
+
+	if diags := utils.CheckError(res, fmt.Sprintf("Unable to get requested index: %s", name)); diags.HasError() {
+		return nil, diags
+	}
+
+	indices := make(map[string]models.Index)
+	if err := json.NewDecoder(res.Body).Decode(&indices); err != nil {
+		return nil, diag.FromErr(err)
+	}
+	index := indices[name]
+	return &index, diags
+}
+
+func (a *ApiClient) DeleteElasticsearchIndexAlias(index string, aliases []string) diag.Diagnostics {
+	var diags diag.Diagnostics
+	log.Printf("[TRACE] Deleting aliases for index %s: %v", index, aliases)
+	res, err := a.es.Indices.DeleteAlias([]string{index}, aliases)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	defer res.Body.Close()
+	if diags := utils.CheckError(res, fmt.Sprintf("Unable to delete aliases '%v' for index '%s'", index, aliases)); diags.HasError() {
+		return diags
+	}
+	return diags
```

Comment:
I'm not proficient at golang, but is there some difference between this and something like this, where you only `return diags` once?

```go
if diags:= utils.CheckError(...) {
  diags.HasError()
}
return diags
```

Reply:
Thanks for pointing this out. Yeah, it can be done better. I will refactor this file a little bit once this PR is merged. There are already lots of changes here to review 😄

Reply:
I'm probably a bit tainted from most of my Golang exposure being the Kubernetes code base but I actually like how this is right now. It is very explicit that if there is an error we want to stop and return immediately. Right now it is safe to let this happen with the return at the end. But future changes could introduce bugs if anything added after this statement won't work in the situation where there was an error earlier.

https://jkl.gg/b/space-shuttle-style-programming/ for context